### ISSUE TITLE
Reverse Face Culling Not Implemented

### ISSUE THREAD

**Issue by mrjustaguy** [godotengine/godot#50005]
### Godot version

v4.0.dev.custom_build [7d88c1a87]

### System information

Windows 10, GLES 2&3, Nvidia Quadro K2000 461.92

### Issue description

Reverse Face Culling doesn't do anything

### Steps to reproduce

Make a scene with Shadows in 3.3.2 stable, toggle Reverse Face Culling, observe changes, then go do the same on current Master branch, observe lack of any changes

### Minimal reproduction project

[Culling.zip](https://github.com/godotengine/godot/files/6734750/Culling.zip)

**Comment by Calinou** [comment]
This may be intended. Now that shadow pancaking has been implemented in the `master` branch, there's little point to using Reverse Cull Face anymore (which turns shadow peter-panning into light bleeding).

**Comment by mrjustaguy** [comment]
Actually, Reverse Cull Face helps with reducing visible Shadow Acne with very low biases, as the Shadow acne appears in the parts that are occluded by the mesh and is generally far easier to get looking great with as little little artifacting visible to the end user as possible. Here's an example
RCF On
![RCF On](https://user-images.githubusercontent.com/50302479/123847363-cefb0f80-d916-11eb-9862-ac962f3cc446.PNG)
RCF Off
![RCF Off](https://user-images.githubusercontent.com/50302479/123847384-d3bfc380-d916-11eb-9422-26a9583832ac.PNG)

**Comment by mrjustaguy** [comment]
Ok Correction, just tested it with Omni Lights, and it also removes some other artifacting that plagues Omni lights at high radiuses with shadows, just have to give a slight negative bias to prevent visible light line around a shadowed object on the floor.

**Comment by mrjustaguy** [comment]
Here all lights in GLES 3:
RCF On
![RCF On all](https://user-images.githubusercontent.com/50302479/123851160-2b602e00-d91b-11eb-8bcd-fa9daa105b2a.PNG)
RCF Off
![Screenshot 2021-06-29 204615](https://user-images.githubusercontent.com/50302479/123851181-31560f00-d91b-11eb-9db4-5f580b78a33e.png)

**Comment by Calinou** [comment]
Do not use a bias value of 0 when Reverse Cull Face is enabled.

**Comment by mrjustaguy** [comment]
Yeah this Bias 0 is not how to use it with RCF disabled, but these same artifacts do appear often even when tweaking stuff around, and worst of all, might not be visible while you're looking close to the actual light source, and appears further away from it, or in the case of Directional shadows, it appears in farther splits.

**Comment by Calinou** [comment]
> Conclusion: Removing RCF Would be bad, as it's a very useful option, and should be considered for being on by default, for a better experience, and Documented better.

As I said, I'm not advocating for Reverse Cull Face to be removed in `3.x`. I'm just saying it doesn't make much sense in `master` because shadow pancaking makes the issue hardly a problem in the first place.

**Comment by mrjustaguy** [comment]
While shadows have improved significantly in master, I disagree with that statement, especially for directional lights.
Directional Lights are really tough to pull off right, because they can very easily get Shadow Acne in the farther away splits, while being totally fine close up. Also Shadow Acne increases as Shadow resolution Decreases, which totally helps with ending up with Shadow Acne on your distant splits, and if a person is testing at High Shadow res, like 16k, as "high" shadow res, and doesn't do all the testing done for high on medium (8k) and low (4k) shadow res, users using said lower resolutions could have a significantly poorer experience due to Shadow Acne issues.

While Reverse Cull Face Might not be a Must have & Must use in 4.0, it doesn't mean it isn't useful, and that it should be removed, as the ease of use with it is undeniable, because instead of having to worry a ton about both Shadow Acne and Peter Panning, you instead worry mostly about the light/dark rims (depending on bias) around the shadow caster if the shadow caster is on or intersecting the shadow recipients.

What should be done with 4.0 regarding Reverse Cull Face is improving the documentation around it, explaining how it changes things, and why one should or shouldn't use it
Also as a side note, as bias is now positive only, with Reverse Cull Face it should multiply the bias by -1 to get a negative result from biasing.

**Comment by Calinou** [comment]
> Directional Lights are really tough to pull off right, because they can very easily get Shadow Acne in the farther away splits, while being totally fine close up.

This is what increasing **Bias Split Scale** is for :slightly_smiling_face:

**Comment by mrjustaguy** [comment]
Bias Split Scale isn't in 4.0, it's in 3.3.2, 4.0 Has Pancakes... Blur does help produce shadow acne, and it is also helping with the issue in the distant splits, but the main issue is that distant splits are just such low effective resolution that Shadow Acne just Creeps in real easy.

Without RCF you are balancing Peter Panning with Distant Shadow Acne, and either way you put it you get quite visible issues, be it the Peter Panning in your close up splits, or Shadow Acne far away and it's simply more complicated to get decent results and be sure that they're going to hold up well under many different viewing positions

With RCF no need to worry about Shadow Acne and Peter Panning (aside from the slight stuff visible close up, the rims between Shadow caster and shadowed mesh) and you're free to play toy around with all the different split locations, resolutions and stuff, and if you manage to get things looking artifact-less close up, it's almost guaranteed that far away things will still look great.

**Comment by Calinou** [comment]
[reduz said that Reverse Cull Face is intended to be supported in `master`:](https://chat.godotengine.org/channel/rendering?msg=xd7TQQcMsga9nHmRe)

```js
<reduz> @Calinou it is not intended though, its likely a bug, there are cases where you still want to reverse it
<reduz> or make it double sided
<Calinou> I guess double-sided makes sense, but this is implemented on a per-mesh basis rather than a per-light basis (in 3.x at least)
<reduz> @Calinou it kinda may make sense per light in some cases too
```

**Comment by markusneg** [comment]
Just wanted to leave another use case in which revert cull faces are almost a Must-have: A 2D top-down scene with 3D geometries and shadows. Even with pancaking and well tuned bias settings, thin structures will leave a visible gap on the ground due to peter panning. With reverse cull face enabled, zero bias settings and ultra-precise shadows are possible because the shadow acne appears on the never visible back of the geometries. Of course this is only possible if no self-shadowing is needed (e.g. a terrain topology casting shadows onto itself).
